AF-2767: Kafka dataset provider

jesuino opened the pull request:
### Kafka Data Set Support for Dashbuilder

Ticket: https://issues.redhat.com/browse/AF-2767

This is a data set to retrieve real time metrics from Kafka. 

## Kafka Data sets

When creating data sets bear in mind that we need to define what kind of service we will be monitoring:

* BROKER: When using Broker you will be monitoring kafka server metrics, general metrics;
* CONSUMER: To monitor consumers and messages handled by a consumer;
* PRODUCER:  To monitor producers and messages created by a producer;

You can also filter the metrics you want to see using the Filter field. It filters by metrics name, type or domain. Attribute is not filtered.

The other fields are related to consumer and producer only, hence they are enabled only when producer/consumer is selected.

Client ID: The client Id - it is required for consumer/producer metrics;
Node ID: The node Id to monitor metrics for a specific node. When filled, topic is ignored
Topic: if you want metrics for a specific topic.

To test this dataset we must set up kafka with one node and a single topic.

## Testing

### Kafka Setup

Kafka Setup, based on https://kafka-tutorials.confluent.io/kafka-console-consumer-producer-basics/kafka.html

1) Download Kafka
2) Start Zookeeper in a console

```
./bin/zookeeper-server-start.sh config/zookeeper.properties
```

3) In a new console start Kafka (Broker) - before export JMX_PORT=9999
```
export JMX_PORT=9999
./bin/kafka-server-start.sh config/server.properties
```

4) Create the topic using some console window
```
./bin/kafka-topics.sh --create --topic example-topic --bootstrap-server localhost:9092 --replication-factor 1 --partitions 1
```
5) In a new console start a Consumer - export JMX_PORT=9998
```
export JMX_PORT=9998
./bin/kafka-console-consumer.sh --group dashbuilder_test --topic example-topic --bootstrap-server localhost:9092
```
6) In a new console start the Producer
```
export JMX_PORT=9997
./bin/kafka-console-producer.sh --topic example-topic --broker-list localhost:9092
```
In the producer window you can type text that will be received by the consumer

![image](https://user-images.githubusercontent.com/359820/108431659-ef58e600-7221-11eb-809d-ce0a3aa19daf.png)


### Data sets on Dashbuilder/BC

Now you can go to `BC -> Admin -> Data sets -> New Data Set` and you will notice a new type for Kafka:

![image](https://user-images.githubusercontent.com/359820/108430502-38a83600-7220-11eb-8ad0-fc67d791fba8.png)

Then you can create datasets for each parts of our Kafka setup using the following parameters:

Host: localhost
Port: 9999 for broker - 9998 for consumer and 9997 for producer

Client ID: console-producer for producer or consumer-dashbuilder_test-1 for consumer
Topic: example-topic

Examples:

BROKER:
![image](https://user-images.githubusercontent.com/359820/108431894-4959ab80-7222-11eb-87d4-4227e2e3dae9.png)

BROKER OUTPUT:
![image](https://user-images.githubusercontent.com/359820/108431926-537baa00-7222-11eb-8e3e-5d6f2db6a374.png)

CONSUMER:
![image](https://user-images.githubusercontent.com/359820/108432052-8625a280-7222-11eb-835a-bb05654d8e6b.png)

CONSUMER OUTPUT
![image](https://user-images.githubusercontent.com/359820/108432068-8de54700-7222-11eb-8312-400aa545d65c.png)


PRODUCER:
![image](https://user-images.githubusercontent.com/359820/108432103-9d649000-7222-11eb-97c7-c4195038f622.png)

PRODUCER OUTPUT:
![image](https://user-images.githubusercontent.com/359820/108432121-a3f30780-7222-11eb-9631-332d8f4a27e5.png)

These 3 datasets can also be found in the following dashboard that can be imported into BC/Dashbuilder:

https://github.com/jesuino/dashbuilder-data/raw/master/dashboards/kafka_simple_dashboard.zip

You will basically also find three simple pages:

![kafka_dashboard](https://user-images.githubusercontent.com/359820/108520346-553c8080-72a9-11eb-8839-6492a7c7c4f7.gif)

Merge with:

https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/1592
kiegroup/kie-soup#177
https://github.com/kiegroup/kie-wb-distributions/pull/1087

jesuino commented:
jenkins execute fdb

jesuino commented:
Hello @caponetto , @sthundat 

Would you please review this PR?

Thanks!

jesuino commented:
jenkins execute fdb

sthundat commented:
jenkins execute fdb

caponetto reviewed (commented):
Hi @jesuino, thanks for the PR! Very interesting stuff.
The code looks great to me. I've only left a few minor comments.
I still want to give it a try, can you share a war file for BC?
And I'll review the `kie-soup` PR as soon as possible.

jesuino commented:
jenkins execute fdb

jesuino commented:
Hello @caponetto 

I made the requested changes, thanks!

I am building a local business-central-webapp because Jenkins FDB is failing to build. It may be due a test failure from this PR, so I will fire another FDB, but in any case I will share a WAR tomorrow.

Thanks!

jesuino commented:
jenkins execute fdb

jesuino commented:
Here's a link to a Business Central Web App that I built locally. Deploy it on Wildfly 19.0.1 - a suggestion is use the jbpm server distribution (7.50.0.FInal) and just replace business-central.war 

https://drive.google.com/file/d/1XHQGowI_-1aYAl9QjUeSS7uvMGPV5Q5Q/view?usp=sharing

caponetto commented:
Hi @jesuino, here are two observations after doing some tests:

1. Should this `size must be between 4 and ...` validation for `port` be there?
![tooltip](https://user-images.githubusercontent.com/638737/109009474-57c61e00-768d-11eb-87be-7667fcce7643.png)

2. If the connection can't be established, can we have a more friendly message? (e.g. in this case I've put a wrong port)
![popup](https://user-images.githubusercontent.com/638737/109009483-5ac10e80-768d-11eb-933c-b8beb63de096.png)

Other than this, pretty nice stuff. :)

jesuino commented:
Hello @caponetto 

Thank you so much for verifying the WAR. 


1) I changed the validation to be more close to a port validation. Now noticed that the message is not correctly encoded - this is not related to this PR, but simply how the hint is built by datasets editors - I can revisit later.

![image](https://user-images.githubusercontent.com/359820/109044361-18112d80-76b1-11eb-9bc4-72c04f40a5a4.png)

2) I did changed the message that is thrown from the Provider, but this is also something at the editors level: it, for some reason, shows the root cause message, which prevents me from adding a human friendly message. This behaviour can also be changed.

The reason I don't do all these improvements is because we were discussing about a new data set editor perspective, which would be a new view created from scratch. But we may have a bulk of improvements for this old data sets editors when we have some time to work on it!

caponetto commented:
Great, @jesuino! It makes sense to me. :)

jesuino commented:
I changed the message to avoid the encoding issue:

![image](https://user-images.githubusercontent.com/359820/109047367-8d323200-76b4-11eb-98d4-52b14d2ba0c5.png)

The commit is in kie-soup: https://github.com/kiegroup/kie-soup/pull/177/commits/6ef072bfca64db4f3968676d4be7327c028d4501

caponetto commented on dashbuilder/dashbuilder-runtime/src/main/webapp/WEB-INF/jboss-deployment-structure.xml:
```diff
@@ -36,6 +36,8 @@
 			<module name="org.jboss.resteasy.resteasy-multipart-provider" />
 			<module name="org.apache.xerces" />
 			<module name="org.apache.xalan" />
+		    <!-- For Kafka Connector -->
+            <module name="org.jboss.remoting-jmx"/>
```
One last thing: this indentation should be aligned with line 38. 😛

sthundat commented:
Hello @jesuino ,
Thanks for the PR.
Below are  some of the observation on kafka dataset,
1.if I select wrong port (port of consumer/producer)for broker not getting connection refuse exception in UI.if i select some other values(ex:9992) connection refused exception is throwing as expected.
![brokerwrongport](https://user-images.githubusercontent.com/56147477/109164046-fa6ec180-779f-11eb-993d-ebbcc147d46a.png)
wrong port selection of broker result in below page
![brokerwrongportdetailswithnodata](https://user-images.githubusercontent.com/56147477/109164139-196d5380-77a0-11eb-9bea-c67932e5095e.png)
2.Is it possible to remove target field overlay information that is displaying just above the drop down.This sometimes preventing users to select the target value from the dropdown.
please note that these are my observation as per current testing.will continue testing next day also.
Thanks,
Savitha

jesuino commented:
Hello @sthundat,

Thanks for your review! Here are my answers for your inquires:

1) That's unfortunately expected. The way the data set works is by scanning MBeans for a specific target: if you select Broker it will scan all Broker MBeans in the provided server. If there's no MBean then it returns an empty data set. It will throw exception only if the server is not available!

2) Could you please share a screenshot or guide me to reproduce this behavior? I know that there's a bug with editors that unfortunately do not select the values in drop downs (it happens also with Remote Server), but it would not be related to this specific provider.

Looking forward for your next reviews, thanks!

sthundat commented:
Hi @jesuino ,
Thanks for your inputs.
Regarding second point,Please find the video below.As you said this is an issue with other dataset as well.
![dropdownissue](https://user-images.githubusercontent.com/56147477/109671622-3e940480-7b9a-11eb-9747-3bce20044a45.gif)
Also if client id field (producer/consumer) is empty user is prompted with exception on screen.But providing wrong client id/topic result in empty dataset.Is it possible to validate this and gives user some information on  what is wrong on input?
Thanks,
Savitha

jesuino commented:
Hello @sthundat 

I made the requested changes:

1) Changed the tooltip to show on top instead on bottom

![image](https://user-images.githubusercontent.com/359820/109851529-8cf4e200-7c32-11eb-9e18-404107b8f080.png)

2) Added an error for when there's no metric (the change is in PR for kie soup https://github.com/kiegroup/kie-soup/pull/177/commits/17b4635e0988ce9d8fc4f522442901144eb8610a):

![image](https://user-images.githubusercontent.com/359820/109851449-7484c780-7c32-11eb-9263-40c191c39d4a.png)

I also changed the root cause of exception in dataset provider, hiding the root cause from the exception threw by the provider, but logging the root cause in DEBUG mode)

![image](https://user-images.githubusercontent.com/359820/109851311-4c956400-7c32-11eb-982e-8e8f49b3467b.png)

Finally I changed validation for host, it may contain less characters

jesuino commented:
jenkins execute fdb

sthundat commented:
jenkins execute fdb

jesuino commented:
Hello,

I built a WAR with the latest changes and made it available here

https://drive.google.com/file/d/1XHQGowI_-1aYAl9QjUeSS7uvMGPV5Q5Q/view?usp=sharing

Thanks!

sthundat commented:
Hi @jesuino ,
Thanks for the changes.Now the validation looks good with the new error messages.
Approving the PR.
Thanks,
Savitha

jesuino commented:
jenkins execute fdb

jesuino commented:
jenkins execute cdb

jesuino commented:
There are multiple test failures, but none of them are related to Dashbuilder

https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/KIE/job/master/job/fdb/job/appformer-master.fdb/136/#showFailuresLink

Will fire a new build to see if the errors were corrected

jesuino commented:
jenkins execute fdb

jesuino commented:
jenkins execute cdb

jesuino commented:
jenkins execute cdb

jesuino commented:
The failed test is unrelated to this PR


![image](https://user-images.githubusercontent.com/359820/110788129-000ee180-824d-11eb-8b5c-96beae15e86f.png)

jesuino commented:
jenkins execute cdb

jesuino commented:
jenkins execute fdb

jesuino commented:
jenkins execute fdb

jesuino commented:
jenkins execute cdb

jesuino commented:
jenkins execute fdb

jesuino commented:
jenkins execute cdb

jesuino commented:
jenkins execute cdb

jesuino commented:
jenkins execute cdb

jesuino commented:
jenkins execute fdb

jesuino commented:
jenkins execute cdb

jesuino commented:
jenkins execute fdb

jesuino commented:
Business Central was finally built with this PR - I downloaded and test locally! IMO we are safe to merge. The failing test is unrelated

![image](https://user-images.githubusercontent.com/359820/111800939-37176f80-88ab-11eb-9419-fea825ee3184.png)

ederign reviewed (approved):
LGTM!
